tasks/push-snapshot/push-snapshot.yaml

```diff
@@ -89,24 +121,85 @@ spec:
+            "$imageTags" '$defaults? + $imageTags? | unique')
+          tags=$(translate_tags "${allTagsPreSubstitution}" "${timestamp}" "${git_sha}" \
+            "${git_sha:0:7}" "${build_sha}")
+
           origin_digest=$(skopeo inspect \
             --override-arch "${arch}" \
             --no-tags \
             --format '{{.Digest}}' \
             "docker://${containerImage}" 2>/dev/null)
 
+          # Push source container if the component has pushSourceContainer: true or if the
+          # pushSourceContainer key is missing from the component and the defaults has
+          # pushSourceContainer: true
+          if [[ $(jq -r '.pushSourceContainer' <<< $component) == "true" ]] \
+            || [[ $(jq 'has("pushSourceContainer")' <<< $component) == "false" && \
+                  ${defaultPushSourceContainer} == "true" ]] ; then
+
+            source_repo=${containerImage%%@sha256:*}
+            source_tag=${origin_digest/:/-}.src
+            # Calculate the source container image based on the provided container image
+            sourceContainer="${source_repo}:${source_tag}"
+            # Check if the source container exists
+            source_container_digest=$(skopeo inspect \
+              --override-arch "${arch}" \
+              --no-tags \
+              --format '{{.Digest}}' \
+              "docker://${sourceContainer}" 2>/dev/null)
+
+            if [ -z "$source_container_digest" ] ; then
+              echo "Error: Source container ${sourceContainer} not found!"
+              exit 1
+            fi
+            # Push the source image with the source tag here. The source image will be
+            # pushed with the provided tags below in the loop
+            push_image "${source_container_digest}" "${name}" "${sourceContainer}" \
+              "${repository}" "${source_tag}" "${arch}"
+          fi
+
+          if [[ $(jq 'length' <<< $tags) -ne 0 ]] ; then
+            commonTags="" # To move out of if statement in v5.0.0
+            for tag in $(jq -r '.[]' <<< $tags) ; do
+              commonTags="${commonTags}${tag} "
+              # Push the container image
+              push_image "${origin_digest}" "${name}" "${containerImage}" "${repository}" "${tag}" \
+              "${arch}"
+
+              # This variable will only exist if the above logic determined the source container should
+              # be pushed for this component
+              if [ -n "${source_container_digest-}" ] ; then
+                push_image "${source_container_digest}" "${name}" "${sourceContainer}" \
+                  "${repository}" "${tag}-source" "${arch}"
+              fi
+            done
+            commonTags=${commonTags% }
+            echo -n $commonTags > $(results.commonTags.path) # To omve out of if in v5.0.0
+            continue
+          fi # all code below this line will be removed once all teams are using the new tag format
```

Comment:
Why not move the comment to its own line? It seems weird to have it next to the `fi` command.

Also, I wonder if we should still have some of these checks somewhere. Like:
```
printf 'Asked to create git sha based tag, but no git sha found in %s\n' "${component}"
```
Or can we just assume that the values will always be populated?

Reply:
I think we can assume the values will always be populated. If users are getting a lot of `null`s in their tags and bringing it to us, we can reassess

Reply:
comment moved

Reply:
ok, sounds good.